Implement transaction fee collection

omerfirmak opened the pull request:
Fixes: #1435

omerfirmak committed:
Transaction: Add Coinbase Transaction type

This will be used for paying fees to the Validators of a Block

CoinBase TXs should never be gossiped or created by the wallets. They are only
created by Validators while nominating transaction sets.

omerfirmak committed:
ValidatorSet: Provide an interface to query about-to-expire Validators

omerfirmak committed:
Fee: Rename DataPayloadChecker to FeeManager

FeeManager will implement all fee related logic. Not just data payload
related stuff.

Also remove the need to include an output to commons budget for
tx with data. Implicit fees are just fine. (sum(input) > sum(output))

Rename getFee to getDataFee to better represent the functionality

Move FeeManager to consensus subdirectory as fees are criticial to reaching consensus.

Add TxBuilder.deduct() to easily simulate fees in TXs

omerfirmak committed:
Fee: Implement fee distribution routines

Amount: Add gcd
Will be used when distributing fees among Validators with
different stakes

Amount: Add percentage() & mul() & count()
Useful when calculating fees

ValidatorSet: Add getValidatorStakes()

omerfirmak committed:
Ledger: Pay fees of Commons Budget

For every block, the required fee amount would be
paid to CommonsBudgetAddress with a single CoinBase transaction

Add note for always checking fees last in Transaction validation

omerfirmak committed:
Fee: Accumulate fees of multiple blocks

The payout to Validators will be periodic, we should accumulate
fees of multiple blocks to be paid in a single Coinbase TX.

omerfirmak committed:
Ledger: Pay the Validators periodically

With each new block, TX fees that will be paid to every Validator is
stored and periodically, a Coinbase TX for each Validator will be created
to distribute the accumulated fees.

Exclude MPVs from fee distribution

omerfirmak committed:
FeeManager: Use ConsensusParams for configuration

AndrejMitrovic commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -132,13 +132,25 @@ public string isInvalidReason (
+            return "Transaction: Coinbase transactions must" ~
+                "include a single Input";
+
+        if (tx.inputs[0] != Input(Hash.init, height))
```
I'm not sure this is the best idea. Couldn't this end up in a situation where there are multiple coinbase transactions (in different blocks) with the same hash? That could end up overwriting a previous utxo.

omerfirmak replied:
I did this to avoid exactly that actually :) Bitcoin had that problem in early days apparently, see BIP34.

> there are multiple coinbase transactions (in different blocks) with the same hash

Wouldnt that be a hash collision anyways? Their inputs will have different heights, shouldnt they have different hashes? As long as we create single coinbase tx per address in a block, we should be fine IMO.

AndrejMitrovic replied:
I don't think the height is part of the tx hash.

AndrejMitrovic replied:
You could do a simple test. Create two exactly same coinbase txs at different heights and see if they produce unique utxos.

omerfirmak replied:
Height actually goes into UTXO hash of the input.

I have tried before, but I'll add an explicit test for it anyways.

AndrejMitrovic replied:
> Height actually goes into UTXO hash of the input.

That doesn't sound right to me. Maybe we're talking about different topics?

AndrejMitrovic replied:
> I did this to avoid exactly that actually :) Bitcoin had that problem in early days apparently, see BIP34.

Yes but the way they did this is to actually include the height in the (hashing) of the coinbase transaction. But I don't see that change in your code?

omerfirmak replied:
The ctor I use there gets a TX hash and output index to build an input. It builds UTXO hash with hash(TX hash, output index). I just repurposed it, the Input it creates has the utxo hash of hash(Hash.init, height). It is just a unique value for each height, but an invalid UTXO hash.

AndrejMitrovic replied:
Oh, it was a sneaky change. Could we add an explicit overload of the ctor that takes a `Height`?

AndrejMitrovic commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -736,6 +742,47 @@ public class ValidatorSet
+    ***************************************************************************/
+
+    public ExpiringValidator[] getExpiringValidators (Height height,
+        out ExpiringValidator[] ex_validators)
```
Needs to be `ref`. `out` sets it to `null`, so setting the `length` later doesn't have the intended effect. Should also call `assumeSafeAppend` after the length is changed to zero.

AndrejMitrovic commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -736,6 +742,47 @@ public class ValidatorSet
+
+            foreach (row; results)
+            {
+                ex_validators ~= ExpiringValidator(Height(row.peek!(size_t)(0)),
```
`size_t` => `ulong`

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1129,6 +1129,25 @@ public class EnrollmentManager
     {
```
commit message: `Add get ValidatorStakes` => `Add getValidatorStakes`

AndrejMitrovic commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -783,6 +783,41 @@ public class ValidatorSet
+
+    ***************************************************************************/
+
+    public UTXO[] getValidatorStakes (UTXOFinder peekUTXO, out UTXO[] utxos)
```
ditto comment about `ref` and `assumeSafeAppend`. And I suggest changing it in the EnrollmentManager too.

AndrejMitrovic commented on source/agora/node/Fee.d:
```diff
@@ -229,6 +236,84 @@ public class FeeManager
+        UTXO[] stakes) nothrow @safe
+    {
+        // no stakes, no fees
+        if (stakes.length == 0)
```
Technically I think this should be an assertion. There should always be active validators.

AndrejMitrovic commented:
The only part that I find tricky here is the use of floating point. This might (probably will) have platform-specific behavior.

omerfirmak commented:
> The only part that I find tricky here is the use of floating point. This might (probably will) have platform-specific behavior.

If you mean the `Amount.percentage()`, I intend to change that to integers only. I was just not sure if it would overflow at that time.

omerfirmak committed:
Ledger&FeeManager: Test fee distribution

omerfirmak committed:
Validation: Validate Coinbase TXs of a Block

Overload appendCoinbaseTX() for convenience

omerfirmak committed:
Test: Add network tests for fee distribution

omerfirmak commented:
I'll add more tests if needed according to the coverage report

AndrejMitrovic commented:
It seems like some of these commits should be squashed together. In particular it doesn't really make sense to introduce functionality in one commit and then add attributes to it in another commit in the very same PR. It seems like a lot of the commits are redesigns of each other. We don't need to keep that history in git.

AndrejMitrovic commented:
I'll review once the logical squashing is done (I don't expect 1 giant commit but just squash the relevant commits together please)

omerfirmak committed:
Ledger: Validate ConsensusData against expected Coinbase TXs

omerfirmak commented:
> I'll review once the logical squashing is done (I don't expect 1 giant commit but just squash the relevant commits together please)

Phew, this took me longer than I'd like to admit. I aged 5 years from the rebase stress :D
Hope it's now easier to review!

omerfirmak commented:
It is weird, CircleCI env is missing https://github.com/dlang/druntime/pull/3012

Geod24 commented:
Unfortunately we are *still* using LDC v1.20.0...

AndrejMitrovic commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -95,8 +95,8 @@ public string isInvalidReason (
-            if (utxo_value.type != TxType.Payment)
-                return "Transaction: Can only freeze a Payment transaction";
+            if (utxo_value.type == TxType.Freeze)
+                return "Transaction: Can't freeze a Freeze transaction";
```
Slightly better wording: `Can't freeze an already frozen transaction`

AndrejMitrovic commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -132,13 +132,25 @@ public string isInvalidReason (
+                "include a single Input";
+
+        if (tx.inputs[0] != Input(height))
+            return "Transaction: Coinbase transactions invalid Input";
```
maybe "Coinbase transaction contains invalid input"

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -358,10 +358,10 @@ public class Ledger
             expected_height, &this.payload_checker.check);
 
-        if (reason !is null || !this.pool.add(tx))
+        if (tx.type == TxType.Coinbase || reason !is null || !this.pool.add(tx))
```
It would be better to skip it before tx validation. Otherwise this might become a DDos vector.

AndrejMitrovic commented on source/agora/utils/Test.d:
```diff
@@ -585,6 +585,29 @@ public struct TxBuilder
+    ***************************************************************************/
+
+    public ref typeof(this) deduct (Amount amount)
+        return
```
This wasn't mentioned in the commit message I think.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -193,6 +193,8 @@ public class Ledger
                             "A block loaded from disk is invalid: " ~
                             fail_reason);
                 }
+                if (block.header.height)
```
I would make this explicit.

AndrejMitrovic commented on source/agora/common/Config.d:
```diff
@@ -174,6 +174,13 @@ public struct NodeConfig
     public uint tx_payload_fee_factor = 200;
+
+    /// The share that Validators would get out of the transction fees
+    /// Out of 100
```
The rest goes to commons budget? I would add that comment.

AndrejMitrovic commented on source/agora/common/Config.d:
```diff
@@ -174,6 +174,13 @@ public struct NodeConfig
+    /// Out of 100
+    public ubyte validator_tx_fee_cut = 70;
+
+    /// How frequent the payments to Validators will be
```
should mention this is the number of blocks (?). Is 144 from the whitepaper?

omerfirmak replied:
144 was @Geod24's suggestion.

AndrejMitrovic commented on source/agora/consensus/validation/Block.d:
```diff
@@ -130,6 +132,26 @@ public string isInvalidReason (const ref Block block, Height prev_height,
+
+    appendCoinbaseTX(expected_txs, block.header.missing_validators);
+
+    if (!isPermutation(expected_txs, block.txs))
```
What does this do?

AndrejMitrovic replied:
It seems like validation becomes memory intensive now.

omerfirmak replied:
Well, I am definitely not proud of this part. I'll try to rework it a little.

omerfirmak replied:
Please check the latest commit, I think it is much better. I will update the docs and squash it, if it gets the OK.

AndrejMitrovic commented on source/agora/consensus/validation/Block.d:
```diff
@@ -139,25 +141,12 @@ public string isInvalidReason (const ref Block block, Height prev_height,
+        block.header.missing_validators);
+    auto incoming_cb_txs = block.txs.filter!(
+            tx => tx.type == TxType.Coinbase);
+    if (!isPermutation(expected_cb_txs, incoming_cb_txs))
```
I'm still not sure about this design. There can only be one coinbase per block, no? Why are there multiple ones being compared here?

omerfirmak replied:
Incoming block could have more than one, we cant rely on that assumption. It would be validated if it only has the single expected coinbase tx.

omerfirmak replied:
If you ask, why I didnt just check for `length == 1` is to make the check somewhat generic.

AndrejMitrovic replied:
I see. Well I guess it's less lines of code, although the actual intention is hard to figure out from the call.

The way I would design it is to always force the coinbase tx to be the first one. That does create issues with sorting, but I would basically only sort on `txs[1..$]`.

Anyway your approach works too.

AndrejMitrovic replied:
My solution would probably create issues with the merkle tree. So yours is better.

omerfirmak replied:
I generally avoid stuff that may create edge cases that I can't immediately recognize. The most generalized solution I can think of is what I go with most of the time.

omerfirmak commented:
Squashed the last commit and rebased.

AndrejMitrovic commented on source/agora/common/Amount.d:
```diff
@@ -112,6 +112,24 @@ public struct Amount
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public static Amount gcd (Amount[] ams)
```
`in Amount[]`?

AndrejMitrovic commented on source/agora/consensus/Fee.d:
```diff
@@ -187,6 +189,12 @@ public class FeeManager
+    public immutable ubyte PayoutPeriod = 144;
+
+    /// Total Amount of fees accumulated per address
+    Amount[PublicKey] accumulated_fees;
```
Missing access attribute.

AndrejMitrovic commented on source/agora/consensus/Fee.d:
```diff
@@ -314,6 +322,108 @@ public class FeeManager
+        foreach (idx, stake; stakes)
+        {
+            this.accumulated_fees.update(stake.output.address,
+            { return validator_fees[idx]; },
+            (ref Amount so_far) {
+                so_far.add(validator_fees[idx]);
+                return so_far;
+            });
```
Needs indent here.

AndrejMitrovic commented on source/agora/consensus/Fee.d:
```diff
@@ -314,6 +322,108 @@ public class FeeManager
+        Calculate total fees of a Transaction set
+
+        Params:
+            tx_set = Transaction set
```
`peekUTXO` is not documented.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -882,6 +904,26 @@ public class Ledger
+        this.enroll_man.getValidatorStakes(&this.utxo_set.peekUTXO, stakes);
+        assert(isStrictlyMonotonic(block.header.missing_validators));
+        foreach (idx; block.header.missing_validators.retro)
+            stakes = stakes.remove(idx);
```
Wouldn't it be more efficient to provide these missing validators in the `getValidatorStakes` call?

I guess it doesn't matter much if slashing will be rare though.

omerfirmak replied:
Yep, I'll do that. Performance benefits may be negligible but it tidies up the caller site. Thanks!

AndrejMitrovic reviewed (commented):
Just a few comments. Otherwise LGTM.

AndrejMitrovic commented:
Sorry about the conflict!

omerfirmak commented:
Rebased

omerfirmak committed:
FeeManager: Add non-volatile storage

Accumulated fees are hard to rebuild when node restarts. Keep a
copy of the in-memory accumulated fees in DB and fetch it when needed.

omerfirmak commented:
I have added DB storage for FeeManager. Please take a look at the last commit.

AndrejMitrovic commented on source/agora/consensus/Fee.d:
```diff
@@ -177,16 +182,33 @@ unittest
+        {
+            const key = PublicKey.fromString(row.peek!(char[])(0));
+            const fee = Amount.fromString(row.peek!(char[])(1));
+            accumulated_fees[key] = fee;
```
this.accumulated_fees

AndrejMitrovic commented on source/agora/consensus/Fee.d:
```diff
@@ -343,6 +365,19 @@ public class FeeManager
+
+            try
+            {
+                auto new_fee = this.accumulated_fees[stake.output.address]
```
This will throw an Error if `accumulated_fees` doesn't have this public key. Is there any case where that could happen?

omerfirmak replied:
Nope, L361 prevents that.

AndrejMitrovic replied:
Got it. It didn't show in the diff.

omerfirmak commented:
Rebased

omerfirmak commented:
Rebased again.

omerfirmak commented:
If there is no new comments on this today, I'll rebase and merge.

omerfirmak commented:
@ferencdg , is the CI failure expected?
